pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+                ActivityTrackingStatus::Enabled {
+                    start: current_session_index + 1,
+                    end: new_status_end_session_index,
+                }
+            } else {
+                <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+                ActivityTrackingStatus::Disabled {
+                    end: new_status_end_session_index,
+                }
+            };
+            <CurrentActivityTrackingStatus<T>>::put(new_status.clone());
+            Self::deposit_event(Event::<T>::ActivityTrackingStatusSet { status: new_status });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists) and activity tracking is enabled
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                if let ActivityTrackingStatus::Enabled { start, end: _ } =
+                    <CurrentActivityTrackingStatus<T>>::get()
+                {
+                    if start <= T::CurrentSessionIndex::session_index() {
+                        total_weight += Self::on_author_noted(orchestrator_chain_author);
+                    }
+                }
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn process_ended_session() {
+            let current_session_index = T::CurrentSessionIndex::session_index();
+            // Since this function will be executed in the beginning of a session
+            // we can populate the active collators from the session that just ended
+            // before resetting the collators storage for the current session
+            // without affecting the current session's active collators records
+            ActiveCollators::<T>::insert(
+                current_session_index.saturating_sub(1),
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            if current_session_index > T::MaxInactiveSessions::get() {
+                <crate::pallet::ActiveCollators<T>>::remove(
+                    current_session_index - T::MaxInactiveSessions::get() - 1,
+                );
+            }
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    if active_collators
+                        .try_insert(author)
```

Comment:
@tmpolaczyk why does try_insert work? according to docs is a nightly feature only: https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.try_insert

Reply:
https://paritytech.github.io/polkadot-sdk/master/sp_core/bounded_btree_map/struct.BoundedBTreeMap.html#method.try_insert

Reply:
thansk!